src/main/java/com/yelp/nrtsearch/server/luceneserver/search/collectors/MyTopSuggestDocsCollector.java

```diff
@@ -79,7 +93,119 @@ public TopSuggestDocs reduce(Collection<TopSuggestDocsCollector> collectors)
+      for (TopSuggestDocs topSuggestDocs : topDocs) {
+        for (TopSuggestDocs.SuggestScoreDoc currentSearchSuggestDoc :
+            topSuggestDocs.scoreLookupDocs()) {
+          int docId = currentSearchSuggestDoc.doc;
+          if (topSuggestScoreDocsUniqueByDoc.containsKey(docId)) {
+            TopSuggestDocs.SuggestScoreDoc existingSearchSuggestDocForDocId =
+                topSuggestScoreDocsUniqueByDoc.get(docId);
+            if (existingSearchSuggestDocForDocId.score < currentSearchSuggestDoc.score) {
+              topSuggestScoreDocsUniqueByDoc.put(docId, currentSearchSuggestDoc);
+            }
+          } else {
+            topSuggestScoreDocsUniqueByDoc.put(docId, currentSearchSuggestDoc);
+          }
+        }
+      }
+
+      // retrieve the top this.numHitsToCollect suggestScoreDocs, sorting them by score
+      TopSuggestDocs.SuggestScoreDoc[] suggestScoreDocs =
+          topSuggestScoreDocsUniqueByDoc.values().stream()
+              .sorted((o1, o2) -> -1 * Float.compare(o1.score, o2.score)) // highest score first
+              .limit(this.numHitsToCollect)
+              .toArray(TopSuggestDocs.SuggestScoreDoc[]::new);
+
+      return new TopSuggestDocs(
+          new TotalHits(this.numHitsToCollect, TotalHits.Relation.EQUAL_TO), suggestScoreDocs);
+    }
+  }
+
+  public static class ExhaustiveTopSuggestDocsCollector extends TopSuggestDocsCollector {
+    private final SuggestScoreDocPriorityQueue priorityQueue;
+
+    public ExhaustiveTopSuggestDocsCollector(int num) {
+      super(num, false);
+      priorityQueue = new SuggestScoreDocPriorityQueue(num);
+    }
+
+    @Override
+    protected void doSetNextReader(LeafReaderContext context) {
+      docBase = context.docBase;
+      // clears priority queue
+      priorityQueue.getResults();
+    }
+
+    @Override
+    public void collect(int docID, CharSequence key, CharSequence context, float score)
+        throws IOException {
+      TopSuggestDocs.SuggestScoreDoc suggestScoreDoc =
+          new TopSuggestDocs.SuggestScoreDoc(docBase + docID, key, context, score);
+      priorityQueue.add(suggestScoreDoc);
+    }
+
+    @Override
+    public TopSuggestDocs get() throws IOException {
+      TopSuggestDocs.SuggestScoreDoc[] results = priorityQueue.getResults();
+      if (results.length > 0) {
+        return new TopSuggestDocs(
+            new TotalHits(results.length, TotalHits.Relation.EQUAL_TO), results);
+      } else {
+        return TopSuggestDocs.EMPTY;
+      }
+    }
+  }
+
+  public static class SuggestScoreDocPriorityQueue {
+    private final Map<Integer, TopSuggestDocs.SuggestScoreDoc> docIdsMap;
+    private final PriorityQueue<TopSuggestDocs.SuggestScoreDoc> priorityQueue;
+
+    public SuggestScoreDocPriorityQueue(int num) {
+      docIdsMap = new HashMap<>(num);
+      priorityQueue = new PriorityQueue<>(num, (o1, o2) -> -1 * Float.compare(o1.score, o2.score));
```

Comment:
It would be good to use the same comparator
```
          (a, b) -> {
            // sort by higher score
            int cmp = Float.compare(b.score, a.score);
            if (cmp == 0) {
              // tie break by completion key
              cmp = Lookup.CHARSEQUENCE_COMPARATOR.compare(a.key, b.key);
              if (cmp == 0) {
                // prefer smaller doc id, in case of a tie
                cmp = Integer.compare(a.doc, b.doc);
              }
            }
            return cmp;
          }
```

Reply:
resolved